### ISSUE TITLE
Support Teradata

### ISSUE THREAD

**Issue by moukat** [metabase/metabase#3102]
JDBC Driver:
http://downloads.teradata.com/download/connectivity/jdbc-driver

Teradata SQL to get list of tables:
SELECT TableName 
FROM dbc.tables 
WHERE DatabaseName='<database name>'
ORDER BY 1

Teradata SQL to get list of table columns:
SELECT ColumnName
FROM DBC.COLUMNS 
WHERE TableName='<table name>' AND DatabaseName = '<database name>'
ORDER BY ColumnId

**Comment by camsaul** [comment]
Hi @moukat, we're adding support for various DBs based on community demand. Since you're the only person who's asked for Teradata thus far, that might be a while 😉 

If you're feeling adventurous you can try [writing a driver yourself](https://github.com/metabase/metabase/wiki/Writing-A-Driver). I'm available to help with any questions you might run into.

**Comment by moukat** [comment]
Hi @camsaul 
I am not familiar at all with Clojure, but some guys in my team are ;-)
few question:
1. How are the JOIN conditions between table defined? based on column name? or should it be manually set?
2. Do you think we can re-use the generic_sql namespace without any change?

**Comment by camsaul** [comment]
Hi @moukat, `JOIN`s are inferred automatically when a query references a column in another table. For SQL drivers, you shouldn't need to worry about that yourself. 

For the most part, your driver can "inherit" from the Generic SQL driver. You do need to supply a some methods to handle things that seem to have no consistent SQL  syntax, like extracting months or years from dates, and you need to do a few other things like supplying a mapping of database types to Metabase internal types.

Take a look at the [Postgres](https://github.com/metabase/metabase/blob/master/src/metabase/driver/postgres.clj) and [MySQL](https://github.com/metabase/metabase/blob/master/src/metabase/driver/mysql.clj) drivers for examples of what you'd be getting yourself into 😉

**Comment by salsakran** [comment]
Also, the single biggest thing to figure out is how to test the driver. 

Will you be able to host an instance we can hit from CI?

**Comment by moukat** [comment]
We developed a Teradata Driver. It's working quite fine. We still need to correct few issues (mainly with time datatypes) to pass the all tests... but our developer is not available anymore to continue this, so if someone is interested to takeover....

**Comment by jburgueno** [comment]
@moukat Hi moukat! would you mind sharing your Teradata Driver implementation here? I'm working on one my self and it would help if you could share what you have.

**Comment by jburgueno** [comment]
@moukat Hi moukat! would you mind sharing your Teradata Driver implementation here? I'm working on one my self and it would help if you could share what you have.

**Comment by moukat** [comment]
Hi @jburgueno, hi all , we could finally finalize this driver and made a pull request here: https://github.com/swisscom-bigdata/metabase/tree/dev/teradata_driver

**Comment by moukat** [comment]
+ Pull Request
https://github.com/metabase/metabase/pull/6459

**Comment by moukat** [comment]
Hi @EmanueleRC , actually we do not provide JAR file. We have just produced the source code and made a pull request. We are now waiting for this pull request to be accepted... When PR accepted, our change will be merged to core project and it should be integrated the official JAR distribution.

Meanwhile, you could
- Vote for this issues in order to show community interest (and share it to all people having potential interest). Add a :+1: on the first comment for that.
- Take our source from our PR and build the JAR yourself (more for testing)

**Comment by pbollu** [comment]
Hi @EmanueleRC,

Metabase is awesome, please add support for Teradata. This will allow Enterprises to try with their existing data in Teradata.

Thank you,
Phani.

**Comment by camsaul** [comment]
Note: open PR is here #6459

**Comment by luca-vercelli** [comment]
any update on this topic?

**Comment by flamber** [comment]
@luca-vercelli Yes, https://github.com/metabase/metabase/pull/6459#issuecomment-471721615

**Comment by Pibbers** [comment]
If you need a Teradata instance for CI, you can use a free instance at this URL https://clearscape.teradata.com/ 

It is a fully compatible Teradata demo environment and accessible via API if necessary.
